Add run_policy to ElasticTask

bgedik opened the pull request:
## Why are the changes needed?

`Elastic` tasks currently do not expose a run policy. As such, the run policy of a `PyTorchJob` created via an `Elastic` task cannot be configured. The same functionality exists for a `PyTorchJob` created via a `PyTorch` task. 

## What changes were proposed in this pull request?

We are adding the same run policy configuration support from `PyTorch` tasks to `Elastic` tasks.

## How was this patch tested?

Unit tests added

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [X] All new and existing tests passed.
- [X] All commits are signed-off.

bgedik commented:
cc @fg91

fg91 commented on plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:
```diff
@@ -137,6 +138,7 @@ class Elastic(object):
     max_restarts: int = 0
     rdzv_configs: Dict[str, Any] = field(default_factory=dict)
+    run_policy: Optional[RunPolicy] = field(default_factory=lambda: None)
 
```
For the `rdzv_configs` which defaults to an empty dict, using a default factory is required to ensure that each instance gets a unique copy of the mutable default value.

I don't think anything speaks against using `run_policy: Optional[RunPolicy] = None` though, or am I overlooking something here?

bgedik replied:
Addressed

fg91 commented on plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:
```diff
@@ -427,6 +429,14 @@ def execute(self, **kwargs) -> Any:
 
         return exception_scopes.user_entry_point(self._execute)(**kwargs)
 
+    def _convert_run_policy(self, run_policy: RunPolicy) -> kubeflow_common.RunPolicy:
```
This method is exactly the same as in `PytorchFunctionTask`. It also doesn't need access to `self`. Could you please remove the duplication by removing the method from the classes and instead turn it into a standalone function called e.g. `_convert_run_policy_to_flyte_idl`?

bgedik replied:
Addressed

fg91 commented on plugins/flytekit-kf-pytorch/tests/test_elastic_task.py:
```diff
@@ -187,3 +188,38 @@ def wf(recoverable: bool):
+    pytorch_tasks = []
+    original_init = PytorchElasticFunctionTask.__init__
+
+    def mock_init(self, *args, **kwargs) -> None:
```
Thanks for adding a test that ensures that the run policy is correctly set by `.get_custom()`.

I find the mocking logic unnecessarily complex. [This](https://github.com/flyteorg/flytekit/blob/2719b349cd6ac9b18cad858ff2fc6d6f79986b1e/plugins/flytekit-kf-pytorch/tests/test_pytorch_task.py#L143) test for the `PytorchFunctionTask` shows how you can call `.get_custom` on a task without mocking the constructor. Can you please simplify it here?

bgedik replied:
Addressed

fg91 reviewed (changes requested):
Thank you for adding the `RunPolicy` to the `Elastic` task config. I tested your PR and can confirm that the `RunPolicy` is correctly applied to the `PytorchJob` object in the cluster and that e.g. a "clean all" run policy works as expected.

I have three small change requests, see comments above. Could you please address them and re-request a review? After that I'll approve 🙏

bgedik commented:
> Thank you for adding the `RunPolicy` to the `Elastic` task config. I tested your PR and can confirm that the `RunPolicy` is correctly applied to the `PytorchJob` object in the cluster and that e.g. a "clean all" run policy works as expected.
> 
> I have three small change requests, see comments above. Could you please address them and re-request a review? After that I'll approve 🙏

Hi @fg91 Addressed your comments

fg91 commented:
> Hi @fg91 Addressed your comments

Can you please sign-off the commits/fix DCO?

bgedik commented:
> > Hi @fg91 Addressed your comments
> 
> Can you please sign-off the commits/fix DCO?

@fg91  Signed
